### ISSUE TITLE
Does Keras support "one class classification" ?

### ISSUE THREAD

**Issue by anujgupta82** [fchollet/keras#1003]
In most simple classification problems, we have two classes: class-0 and class-1. 

In my case I only have the data for Class 0 and none for class 1. I am thinking of making a model based on "One class classification" approach [http://stats.stackexchange.com/questions/76420/classifier-for-only-one-class]

Does current Keras support "One class classification" approach ? If not, any plans of putting it in your roadmap ?

**Comment by lemuriandezapada** [comment]
Just have a network with one output unit, sigmoid activation layer and use binary crossentropy as loss function, and have training labels as 0 for one class and 1 for the other. It's actually easier to implement than the multiclass system.

**Comment by anujgupta82** [comment]
https://github.com/lemuriandezapada: thanks for your response. But you missed the core of my question when you say "and have training labels as 0 for one class and 1 for the other". My training set has no "other" class. The training set consists entirely of positive examples of Class 0. There is no positive examples for Class 1

**Comment by dbonadiman** [comment]
So what you are trying to accomplish is basically to build some kind  of anomaly detection model?

**Comment by anujgupta82** [comment]
https://github.com/dbonadiman: in some sense u can say that. Need for "one class classification" arises often enough - For example Imagine a factory type of setting; heavy machinery under constant surveillance of some advanced system. The task of the controlling system is to determine when something goes wrong; the products are below quality, the machine produces strange vibrations or something like a temperature that rises. It is relatively easy to gather training data of situations that are OK; it is just the normal production situation. But on the other side, collection example data of a faulty system state can be rather expensive, or just impossible. If a faulty system state could be simulated, there is no way to guarantee that all the faulty states are simulated and thus recognized in a traditional two-class problem.

The need for "one class classification" arises when either:
(1) you dont have data for other class. 
(2) or other class is far too varied to make it impossible to collect data comprehensivly for the other class. 

Factory example belongs to (1). For (2) considers tweets on twitter. there are various types of emotions - happy, sad, angry, indifferent, bored, anxious, and so on (huge list). I am interested in indentifying (only) - "if a tweet has happiness enotion or not". Say I am able to collect a corpse of tweets with happy emotion. How does one collect a corpous of rest all emotions ? One-Class Support Vector Machine is a popular idea [http://rvlasveld.github.io/blog/2013/07/12/introduction-to-one-class-support-vector-machines/]

I am looking for a similar idea in neural nets. Hope I made some sense.

**Comment by dbonadiman** [comment]
I don't know if this can lead you somewhere but you can try something with auto-encoders probably. Try to encode-decode your examples in the 0 class. probably you are going to model the zeroes  distribution on your data. You train that on a lot of data. then when a 1 class example arises you will have more reconstruction error. Since your example do not belong to the same distribution.

By the way i cannot help you more, that is an unusual setup for NNs

**Comment by LeavesBreathe** [comment]
@anujgupta82 I apologize if this has been suggested already, but you may want to consider using the sigmoidal cross-entropy. 

Are you sure you can't distort your data to make a class 1? Usually you can apply many types of randomization and entropy techniques to distort data. But I don't want to beat a dead horse here.

**Comment by adywp** [comment]
@anujgupta82 Have you try autoencoder approach? (http://www.sciencedirect.com/science/article/pii/S092523120600261X)
It's a FeedForward NN that try to reconstruct input at the output layer.
I am new to Keras, so can someone give a simple model example in Keras for this autoencoder approach?

**Comment by bandrei** [comment]
Binary classification is different from one-class classification. Conceptually you can think of binary classification as a plane that separates high dimensional data, whereas one-class classification encapsulates the data. If you were to apply similar principals from one-class SVMs to DNN you would need to create a function that minimizes the output with respect to some spatial center for the data (i.e. rather than having a y^hat and y variables you would need to use some "center" and y in your loss function).

I should do some research on this haha.

**Comment by taninaim** [comment]
Read this paper to understand the difference between a binary classification problem and a one-class classification problem https://arxiv.org/pdf/1801.05365.pdf

**Comment by Sorkanius** [comment]
does any of you have a follow up? An alternative to Isolation forest and OC SVM based on NN?

**Comment by wiliyam** [comment]
@sadeghmir

**Comment by nader-chmait** [comment]
Technically your problem is not a classification problem but rather is an object detection problem. There are heaps of documentation about how to implement this online. Hope that helps.

**Comment by Sorkanius** [comment]
No, it is not an object detection problem. It is a one class classification, not even involving images.

**Comment by nader-chmait** [comment]
Thanks for the clarification. In that case it should be possible to generate artificial data and treat the problem as a binary classification problem. For the sake of the example, assume all your attributes are numerical. You can generate new data for the “other” class by sampling from some distribution (e.g. normal random numbers) and evaluating them relative to your actual class. You can use something like a propensity score matching methodology to check how balanced or similar your artificial data is to the original one and discard those observations that are indeed similar. Note that the propensity scores should be calculated on all attributes except your output variable.

**Comment by Sorkanius** [comment]
but then I will introduce bias to my estimator since it will learn to differenciate between my class and the generated data. I do not want that. I want a framework such as the [OC-SVM ](http://users.cecs.anu.edu.au/~williams/papers/P126.pdf) but based on NN in order to have greater approximation capabilities.

**Comment by nader-chmait** [comment]
Thanks for sharing the link. Check my updated answer just in case it helps.

**Comment by Sorkanius** [comment]
Thank you! The key is that I do not have data from other classes. So I need a different approach. Such as this one: https://arxiv.org/abs/1802.04881

**Comment by franva** [comment]
hit the same problem.

Do we have a solution now?

**Comment by Sorkanius** [comment]
Not that I know of. CHeck out OC SVm and Isolation Forest, should work well if you have plain data. With images use and Autoencoder. That's my recommendation

**Comment by franva** [comment]
appreciated

**Comment by pabloacera** [comment]
There are several papers implementing oc-nn with NNs, any intention to include the oc-nn loss in keras?
Thanks!

**Comment by BeniaminK** [comment]
I believe the problem is similar with an "open-set classification", which is sumed-up nicely here: [https://arxiv.org/pdf/1811.08581.pdf](https://arxiv.org/pdf/1811.08581.pdf)

**Comment by molejnik-mergebit** [comment]
I have the same problem - I want to use Keras to build one-class classifier (OC SVM or isolation forest). Is it possible with the current version? BTW, why is this issue marked as "Closed"?
